webui/react/src/components/CreateExperimentModal.tsx

```diff
@@ -0,0 +1,88 @@
+import { Alert, Modal } from 'antd';
+import yaml from 'js-yaml';
+import React, { SetStateAction, useCallback, useState } from 'react';
+import MonacoEditor from 'react-monaco-editor';
+
+import { routeAll } from 'routes';
+import { forkExperiment } from 'services/api';
+
+import css from './CreateExperimentModal.module.scss';
+
+export interface CreateExpModalState {
+  visible: boolean;
+  config: string;
+}
```

Comment:
**suggestion:** these two states don't really have anything in common, and by that I mean that when you call `setState`, you are only really changing one or the other and not the two simultaneously. Rather than bundling them into a separate interface, it'll be easier to handle code that treats them separately.
```
interface Props {
  ...
  config: string;
  onCancel: () => void;
  onConfigChange: () => string;
  visible?: boolean;
}
```

Reply:
there is another concern of passing in a function that can mutate a state of a different component directly that's not itself instead of a buffered event handler.

Reply:
I had it the other way first but here I'm thinking we could separate writeable states of the component into its own group as I'm anticipating we add more when we add the form on top or more features.

update: I separated the read part of the states into separate props but as we talked I left the set part of it.

Reply:
I left out the part that creates another handler on top of the setState to let the interface of the setState which allows both passing the new state and a function to set the state but also lower changes to experiment details page as you already have another or open there.